x/vote/manager/manager.go

```diff
@@ -0,0 +1,294 @@
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return err
+	}
+
+	interest, err := vm.gm.GetInterestSince(ctx, voter.LastPowerChangeAt, voter.LinoStake)
+	if err != nil {
+		return err
+	}
+
+	if err := vm.am.AddCoinToUsername(ctx, username, voter.Interest.Plus(interest)); err != nil {
+		return err
+	}
+	voter.Interest = linotypes.NewCoinFromInt64(0)
+	voter.LastPowerChangeAt = ctx.BlockHeader().Time.Unix()
+	return vm.storage.SetVoter(ctx, username, voter)
+}
+
+func (vm VoteManager) GetVoterDuty(ctx sdk.Context, username linotypes.AccountKey) (types.VoterDuty, sdk.Error) {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return types.DutyVoter, err
+	}
+	return voter.Duty, nil
+}
+
+func (vm VoteManager) GetLinoStake(ctx sdk.Context, username linotypes.AccountKey) (linotypes.Coin, sdk.Error) {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return linotypes.NewCoinFromInt64(0), err
+	}
+	return voter.LinoStake, nil
+}
+
+// AssignDuty froze some amount of stake and assign a duty to user.
+func (vm VoteManager) AssignDuty(
+	ctx sdk.Context, username linotypes.AccountKey, duty types.VoterDuty, frozenAmount linotypes.Coin) sdk.Error {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return err
+	}
+	if voter.Duty != types.DutyVoter {
+		return types.ErrNotAVoterOrHasDuty()
+	}
+
+	if voter.FrozenAmount.IsPositive() {
+		return types.ErrFrozenAmountIsNotEmpty()
+	}
+
+	if !voter.LinoStake.IsGTE(frozenAmount) {
+		return types.ErrInsufficientStake()
+	}
+
+	interest, err := vm.gm.GetInterestSince(ctx, voter.LastPowerChangeAt, voter.LinoStake)
+	if err != nil {
+		return err
+	}
+
+	voter.Interest = voter.Interest.Plus(interest)
+	voter.Duty = duty
+	voter.FrozenAmount = frozenAmount
+	// voter.LinoStake = voter.LinoStake.Minus(frozenAmount)
+	voter.LastPowerChangeAt = ctx.BlockHeader().Time.Unix()
+
+	if err := vm.storage.SetVoter(ctx, username, voter); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UnassignDuty register unassign duty event with time after waitingPeriodSec seconds.
+func (vm VoteManager) UnassignDuty(ctx sdk.Context, username linotypes.AccountKey, waitingPeriodSec int64) sdk.Error {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return err
+	}
+	if voter.Duty == types.DutyVoter {
+		return types.ErrNoDuty()
+	}
+	return vm.gm.RegisterEventAtTime(
+		ctx, ctx.BlockHeader().Time.Unix()+waitingPeriodSec, types.UnassignDutyEvent{Username: username})
+}
+
+// SlashStake - slash as much as it can, regardless of frozen money
+func (vm VoteManager) SlashStake(ctx sdk.Context, username linotypes.AccountKey, amount linotypes.Coin) sdk.Error {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return err
+	}
+
+	// slash amount should be larger than stake
+	// frozen money doesn't affect slash logic
+	if !voter.LinoStake.IsGTE(amount) {
+		return types.ErrInsufficientStake()
```

Comment:
Since this is slash, why not just minus it's stake to zero when no enough stake?